File: grind75/733.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: 
int) -> List[List[int]]:
        # dfs
        stack = []
        stack.append([sr, sc])
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        m = len(image)
        n = len(image[0])
        old_color = image[sr][sc]
        visited = set()
        while stack:
            r, c = stack.pop()
            if r < 0 or r >= m or c < 0 or c >= n:
                continue
            
            if image[r][c] != old_color:
                continue
            
            image[r][c] = color
            for dr, dc in directions:
                new_r = r + dr
                new_c = c + dc
                if (new_r, new_c) not in visited:
                    stack.append([new_r, new_c])
                    visited.add((new_r, new_c))
        
        return image

File: grind75/733.go
func floodFill(image [][]int, sr int, sc int, color int) [][]int {
    dirs := [][]int{[]int{1,0}, []int{-1,0}, []int{0,1}, []int{0,-1},}
    m,n := len(image), len(image[0])
    visited := make([]int, m*n)
    old := image[sr][sc]
    var dfs func(int, int)
    dfs = func(i, j int) {
        for _, dir := range dirs {
            ni, nj := i+dir[0], j+dir[1]
            if ni < 0 || ni >= m || nj < 0 || nj >= n || visited[ni*n+nj] == 1 {
                continue
            }
            if image[ni][nj] != old {
                continue
            }
            image[ni][nj] = color
            visited[ni*n+nj] = 1
            dfs(ni,nj)
        }
    }
    visited[sr*n+sc] = 1
    image[sr][sc] = color
    dfs(sr, sc)
    return image
}
